### PR TITLE
Allow for permanent tracking of compilation times

### PR DESCRIPTION
**Pull request by janrous-rai**
In situations where we run long-running server that spends non-trivial amount of time in the compilation we are happy to pay the cost to have this instrumentation always on.

This PR allows us to enable permanent tracking (effectively switching the before/after switches moot) by calling `jl_track_compile_time_permanently()` method, and then providing `jl_cumulative_compile_time_ns_total()` to return the total time.

### CODE REVIEW

**Comment by janrous-rai** [comment]
fyi @vchuravy, @vtjnash  and @IanButterworth who were reviewing Nathan's original re-entrancy PR

**Review by NHDaly — APPROVED** [review]
I love it! 😁 

At this point, I've written most of it, so probably good to have another set of eyes on it, but yeah, this LGTM!

Thanks for the initial idea, @janrous-rai - i think this is a solid improvement. 👍

**Comment by janrous-rai** [comment]
Looks good, thanks @NHDaly for working through this. Let me know if you would like me to resolve the feedback on this or you consider yourself the new "owner" of this until approved?

Regarding the suggestions about 1. total compile time tracking and 2. per-thread compile time tracking -- I think that (1) is definitely useful and I'm not really sure if (2) has much value. Yes, generally speaking, more metrics is better, but with task migration, threads become even more "opaque" and not really interesting so I would be completely fine with having just the per-process total.

It might still make sense to measure per-thread (or total) compilation contention (waiting on the lock held by other threads) as this likely may result in observable performance loss, but this is for another PR altogether.

File: src/jitlayers.cpp
Comment (NHDaly) [reply src/jitlayers.cpp:79]: Yeah.... i've marked #41923 for `#triage`.

Now that we've talked about it, i think that would definitely be the better approach than this one, yeah.

**Comment by NHDaly** [comment]
@janrous-rai - yeah, i'll take it over. sorry for the delay. I think waiting for #41923 is definitely better, but it'll have to wait until triage next week. Thanks and sorry

**Comment by NHDaly** [comment]
OKAY, so, now that https://github.com/JuliaLang/julia/pull/41923 is merged, i think we can actually close this PR! 😮 

I think users can globabally and permanently enable compilation time tracking by calling `Base.cumulative_compile_time_ns_before()`, and that's it! :)

You don't need to worry about it getting turned off on an exception, because we've handled that with the try-catch, so if you manually bump the start count with `cumulative_compile_time_ns_before()`, it'll never turn off. 😊

This was quite a journey, but I'm glad it's settled.

So I'm going to close this PR now! @janrous-rai let me know if I'm missing anything else!
Thanks @Sacha0 and @vchuravy for the help to get here. 💪

**Pull request closed without merging**
